### PR TITLE
fix: output router capture wrong router logits in qwen moe models

### PR DESCRIPTION
**Pull request by ITcarrot**
# What does this PR do?

This PR fixes a bug in the router implementation of several MoE models (Qwen Moe like models, `Olmoe`, `FlexOlmo`). 

Previously, the raw `router_logits` were being overwritten by the result of the softmax operation:
```python
router_logits = F.linear(hidden_states, self.weight)
router_logits = torch.nn.functional.softmax(router_logits, dtype=torch.float, dim=-1)
```
This behavior destroys the raw logits information, which is often required for:
1. Casting returned `router_top_value` back to input type in `router_top_value = router_top_value.to(router_logits.dtype)`
2. Calculating auxiliary losses (e.g., load balancing loss) which typically operate on raw logits.
3. Returning correct router_logits in the model output if requested.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@Cyrilvallez @IlyasMoutawwakil

### CODE REVIEW

**Review by vasqu — APPROVED** [review]
Good catch yes, the logits should not go through softmax :sweat_smile: 

Could we have a small test added to qwen2 moe (since everyone else inherits from this) + are there any more such cases?
